R/clean_redcap_long.R

```diff
@@ -86,6 +75,49 @@ clean_redcap_long <- function(db_data_long,
     structure = "nonrepeating"
   )
 
+  ## Repeating Instruments Logic ----
+  if (has_repeat_forms) {
+    # If mixed structure allowed, retrieve mixed structure forms
+
+    mixed_structure_ref <- data.frame()
+
+    if (allow_mixed_structure) {
+      # Retrieve mixed structure fields and forms in reference df
+      mixed_structure_ref <- get_mixed_structure_fields(db_data_long) %>%
+        filter(.data$rep_and_nonrep & !str_ends(.data$field_name, "_form_complete")) %>%
+        left_join(db_metadata_long %>% select(.data$field_name, .data$form_name),
+          by = "field_name"
+        )
+
+      # Update if project actually has mixed structure
+      has_mixed_structure_forms <- nrow(mixed_structure_ref) > 0
+      # Convert mixed instruments to mixed structure format
+    } else {
+      if (!has_mixed_structure_forms) {
+        check_repeat_and_nonrepeat(db_data_long)
+      }
+    }
+
+    repeated_forms_tibble <- tibble(
+      redcap_form_name = repeated_forms,
+      redcap_data = map(
+        .data$redcap_form_name,
+        ~ distill_repeat_table_long(
+          .x,
+          db_data_long,
+          db_metadata_long,
+          linked_arms,
+          has_mixed_structure_forms = has_mixed_structure_forms,
```

Comment:
I think this is fine but noting that one implication of this is that `convert_mixed_instrument()` gets run for every instrument even if it's actually repeating rather than mixed. It works because because `mixed_structure_ref` gets filtered to 0 rows for repeating instruments and the `for` loop in `convert_mixed_instrument()` doesn't run. We do still need to `filter()` `mixed_structure_ref` every time to find that out though which could be a performance hit. No need to optimize until it's a problem though

Reply:
Does it? The actual conversion is wrapped in a check on `has_mixed_structure_forms` so if `FALSE` (which it would be for the whole `map()` sequence) then `convert_mixed_instrument()` wouldn't get run unless I'm missing something.

https://github.com/CHOP-CGTInformatics/REDCapTidieR/blob/65d5baf88a5d916e22a385957b7dd5e304409607/R/clean_redcap_long.R#L315-L317

Reply:
Right but if you have mixed and repeating instruments then `has_mixed_structure_forms` is `TRUE` and it gets run for all the repeating instruments in addition to the mixed instruments.